compiler/rustc_codegen_llvm/src/gotoc/mir_to_goto/codegen/statement.rs

```diff
@@ -99,27 +99,74 @@ impl<'tcx> GotocCtx<'tcx> {
-        let instance = Instance::resolve_drop_in_place(self.tcx, loc_ty);
-        if let Some(hk) = self.hooks.hook_applies(self.tcx, instance) {
+        let drop_instance = Instance::resolve_drop_in_place(self.tcx, loc_ty);
+        if let Some(hk) = self.hooks.hook_applies(self.tcx, drop_instance) {
             let le = self.codegen_place(location).goto_expr;
-            hk.handle(self, instance, vec![le], None, Some(*target), None)
+            hk.handle(self, drop_instance, vec![le], None, Some(*target), None)
         } else {
-            let drop_implementation = match instance.def {
-                InstanceDef::Virtual(..) => self
-                    .codegen_unimplemented(
-                        "drop_in_place for dynamic argument",
-                        Type::Empty,
-                        Location::none(),
-                        "https://github.com/model-checking/rmc/issues/11",
-                    )
-                    .as_stmt(Location::none()),
-                InstanceDef::DropGlue(..) => Stmt::skip(Location::none()),
+            let drop_implementation = match drop_instance.def {
+                InstanceDef::DropGlue(_, None) => {
+                    // We can skip empty DropGlue functions
+                    Stmt::skip(Location::none())
+                }
                 _ => {
-                    let func = self.codegen_func_expr(instance, None);
-                    func.call(vec![self.codegen_place(location).goto_expr.address_of()])
-                        .as_stmt(Location::none())
+                    match loc_ty.kind() {
+                        ty::Dynamic(..) => {
+                            // Virtual drop via a vtable lookup
+                            let trait_fat_ptr =
+                                self.codegen_place(location).fat_ptr_goto_expr.unwrap();
+
+                            // Pull the function off of the fat pointer's vtable pointer
+                            let vtable_ref =
+                                trait_fat_ptr.to_owned().member("vtable", &self.symbol_table);
+                            let vtable = vtable_ref.dereference();
+                            let fn_ptr = vtable.member("drop", &self.symbol_table);
+
+                            // Pull the self argument off of the fat pointer's data pointer
+                            let self_ref =
+                                trait_fat_ptr.to_owned().member("data", &self.symbol_table);
+                            let self_ref =
+                                self_ref.cast_to(trait_fat_ptr.typ().clone().to_pointer());
+
+                            let func_exp: Expr = fn_ptr.dereference();
+                            func_exp.call(vec![self_ref]).as_stmt(Location::none())
+                        }
+                        _ => {
+                            // Non-virtual, direct drop call
+                            assert!(!matches!(drop_instance.def, InstanceDef::Virtual(_, _)));
```

Comment:
Can we assert this above? Would that be a better check than matching on `loc_ty.kind()`

Reply:
The conditional doesn't work as a check, because asserting the inverse in the dynamic case fails whenever the instance def is `InstanceDef::DropGlue` instead of `::Virtual`. This structure also matches how the Cranelift backend handles this case.